File: 백준/Gold/1516. 게임 개발/게임 개발.py
from sys import stdin
from collections import defaultdict
from heapq import heappush, heappop
input = stdin.readline

n = int(input())
ans = [0] * (n+1)
time = [0] * (n+1)
indegree = [0] * (n+1)
graph = defaultdict(list)

for i in range(1,n+1):
    inp = list(map(int,input().split()))
    time[i] = inp[0]
    for p in inp[1:-1]:
        indegree[i] += 1
        graph[p].append(i)

hq = []
for i in range(1,n+1):
    if not indegree[i]:
        heappush(hq,(time[i],i))

while hq:
    t, cur = heappop(hq)
    ans[cur] = t
    for nxt in graph[cur]:
        indegree[nxt] -= 1
        if not indegree[nxt]:
            heappush(hq,(t + time[nxt],nxt))

for x in range(1,n+1):
    print(ans[x])

File: 백준/Gold/1516. 게임 개발/게임 개발.java
import java.io.*;
import java.util.*;

public class Main {

	public static void main(String[] args) throws Exception {
		//System.setIn(new FileInputStream("res/input.txt"));
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st;
		StringBuilder sb = new StringBuilder();
		int n = Integer.parseInt(br.readLine());
		int[] record = new int[n + 1];
		int[] time = new int[n + 1];
		int[] indegree = new int[n + 1];
		ArrayList<Integer>[] graph = new ArrayList[n + 1];
		for (int i = 1; i <= n; i++) {
			graph[i] = new ArrayList<>();
		}
		for (int i = 1; i <= n; i++) {
			st = new StringTokenizer(br.readLine());
			time[i] = Integer.parseInt(st.nextToken());
			while (st.hasMoreTokens()) {
				int nxt = Integer.parseInt(st.nextToken());
				if (nxt != -1) {
					graph[nxt].add(i);
					indegree[i]++;
				}
			}
		}
		ArrayDeque<Integer> q = new ArrayDeque<>();
		for (int i = 1; i <= n; i++) {
			if (indegree[i] == 0) {
				q.offerLast(i);
			}
		}
		while (!q.isEmpty()) {
			int cur = q.pollFirst();
			for (int nxt : graph[cur]) {
				indegree[nxt]--;
				record[nxt] = Math.max(record[nxt], record[cur] + time[cur]);
				if (indegree[nxt] == 0) {
					q.offerLast(nxt);
				}
			}
		}
		for (int i = 1; i <= n; i++) {
			sb.append(record[i] + time[i]).append("\n");
		}
		System.out.println(sb);
	}
}